hpl-2.2/man/man3/HPL_rollT.3
.TH HPL_rollT 3 "February 24, 2016" "HPL 2.2" "HPL Library Functions"
.SH NAME
HPL_rollT \- Roll U and forward the column panel.
.SH SYNOPSIS
\fB\&#include "hpl.h"\fR
 
\fB\&void\fR
\fB\&HPL_rollT(\fR
\fB\&HPL_T_panel *\fR
\fI\&PBCST\fR,
\fB\&int *\fR
\fI\&IFLAG\fR,
\fB\&HPL_T_panel *\fR
\fI\&PANEL\fR,
\fB\&const int\fR
\fI\&N\fR,
\fB\&double *\fR
\fI\&U\fR,
\fB\&const int\fR
\fI\&LDU\fR,
\fB\&const int *\fR
\fI\&IPLEN\fR,
\fB\&const int *\fR
\fI\&IPMAP\fR,
\fB\&const int *\fR
\fI\&IPMAPM1\fR
\fB\&);\fR
.SH DESCRIPTION
\fB\&HPL_rollT\fR
rolls the local arrays containing the local pieces of U, so
that on exit to this function  U  is replicated in every process row.
In addition, this function probe for the presence of the column panel
and forwards it when available.
.SH ARGUMENTS
.TP 8
PBCST   (local input/output)    HPL_T_panel *
On entry,  PBCST  points to the data structure containing the
panel (to be broadcast) information.
.TP 8
IFLAG   (local input/output)    int *
On entry, IFLAG  indicates  whether or not  the broadcast has
already been completed.  If not,  probing will occur, and the
outcome will be contained in IFLAG on exit.
.TP 8
PANEL   (local input/output)    HPL_T_panel *
On entry,  PANEL  points to the data structure containing the
panel (to be rolled) information.
.TP 8
N       (local input)           const int
On entry, N specifies the local number of rows of  U.  N must
be at least zero.
.TP 8
U       (local input/output)    double *
On entry,  U  is an array of dimension (LDU,*) containing the
local pieces of U in each process row.
.TP 8
LDU     (local input)           const int
On entry, LDU specifies the local leading dimension of U. LDU
should be at least  MAX(1,N).
.TP 8
IPLEN   (global input)          const int *
On entry, IPLEN is an array of dimension NPROW+1.  This array
is such that IPLEN[i+1] - IPLEN[i] is the number of rows of U
in each process row.
.TP 8
IPMAP   (global input)          const int *
On entry, IMAP  is an array of dimension  NPROW.  This  array
contains  the  logarithmic mapping of the processes. In other
words,  IMAP[myrow]  is the absolute coordinate of the sorted
process.
.TP 8
IPMAPM1 (global input)          const int *
On entry,  IMAPM1  is an array of dimension NPROW. This array
contains  the inverse of the logarithmic mapping contained in
IMAP: For i in [0.. NPROW) IMAPM1[IMAP[i]] = i.
.SH SEE ALSO
.BR HPL_pdlaswp01T \ (3).


hpl-2.2/HISTORY
==============================================================
 High Performance Computing Linpack Benchmark (HPL)
 HPL - 2.2 - February 24, 2016
==============================================================

 History

 - 09/09/00 Public release of Version 1.0

 - 09/27/00 A couple of mistakes in the  VSIPL  port have been
 corrected.  The tar file as well as the web site were updated
 on September 27th, 2000.  Note  that  these problems were not
 affecting the BLAS version of the software in any way.

 - 01/01/04 Version 1.0a
 The  MPI  process grid  numbering  scheme  is now an run-time
 option.
 The inlined assembly  timer  routine that caused the compila-
 tion to fail when using  gcc  version 3.3  and above has been
 removed from the package.
 Various building problems on the T3E have been fixed;  Thanks
 to Edward Anderson.

 - 15/12/04 Version 1.0b
 Weakness of the pseudo-random matrix generator found for pro-
 blem sizes being power of twos and larger  than 2^15;  Thanks
 to Gregory Bauer. This problem has not been fixed. It is thus
 currently recommended to  HPL  users willing to test matrices
 of size larger than 2^15 to not use power twos.

 When the matrix size is such that one needs  > 16 GB  per MPI
 rank,  the  intermediate  calculation  (mat.ld+1) * mat.nq in
 HPL_pdtest.c  ends up  overflowing  because  it is done using
 32-bit arithmetic.  This issue has been fixed by  typecasting
 to size_t; Thanks to John Baron.

 - 09/10/08 Version 2.0

 Piotr Luszczek changed to 64-bit RNG, modified files:
 -- [M] include/hpl_matgen.h
 -- [M] testing/matgen/HPL_ladd.c
 -- [M] testing/matgen/HPL_lmul.c
 -- [M] testing/matgen/HPL_rand.c
 -- [M] testing/ptest/HPL_pdinfo.c

 For a motivation for the change, see:
    Dongarra and Langou, ``The Problem with the Linpack
    Benchmark Matrix Generator'', LAWN 206, June 2008.

 -- [M] testing/ptest/HPL_pdtest.c  --

 Julien Langou changed the test for correctness from
      ||Ax-b||_oo / ( eps * ||A||_1  * N            )
      ||Ax-b||_oo / ( eps * ||A||_1  * ||x||_1      )
      ||Ax-b||_oo / ( eps * ||A||_oo * ||x||_oo * N )
 to the normwise backward error
      || r ||_oo / ( eps * ( || x ||_oo * || A ||_oo + || b ||_oo ) * N )
 See:
  Nicholas J. Higham, ``Accuracy and Stability of Numerical Algorithms'',
  Society for Industrial and Applied Mathematics, Philadelphia, PA, USA,
  Second Edition, pages = xxx+680, ISBN = 0-89871-521-0, 2002.

 Note that in our case || b ||_oo is almost for sure
 1/2, we compute it anyway.

 - 10/26/2012 Version 2.1

 Piotr Luszczek introduced exact time stamping for HPL_pdgesv():
 -- [M] dist/include/hpl_misc.h
 -- [M] dist/testing/ptest/HPL_pdtest.c

 Piotr Luszczek fixed out-of-bounds access in data spreading functions
 and exact time stamping for HPL_pdgesv():
 -- [M] dist/src/pgesv/HPL_spreadN.c
 -- [M] dist/src/pgesv/HPL_spreadT.c
 Thanks to Stephen Whalen from Cray.

 - 02/24/2016 Version 2.2

 Piotr Luszczek added continuous reporting of factorization progress
 submitted by Intel and make scripts that uses Intel software tools and
 libraries and their Apple's Mac OS X equivalents.


hpl-2.2/src/pgesv/HPL_rollT.c
/* 
 * -- High Performance Computing Linpack Benchmark (HPL)                
 *    HPL - 2.2 - February 24, 2016                          
 *    Antoine P. Petitet                                                
 *    University of Tennessee, Knoxville                                
 *    Innovative Computing Laboratory                                 
 *    (C) Copyright 2000-2008 All Rights Reserved                       
 *                                                                      
 * -- Copyright notice and Licensing terms:                             
 *                                                                      
 * Redistribution  and  use in  source and binary forms, with or without
 * modification, are  permitted provided  that the following  conditions
 * are met:                                                             
 *                                                                      
 * 1. Redistributions  of  source  code  must retain the above copyright
 * notice, this list of conditions and the following disclaimer.        
 *                                                                      
 * 2. Redistributions in binary form must reproduce  the above copyright
 * notice, this list of conditions,  and the following disclaimer in the
 * documentation and/or other materials provided with the distribution. 
 *                                                                      
 * 3. All  advertising  materials  mentioning  features  or  use of this
 * software must display the following acknowledgement:                 
 * This  product  includes  software  developed  at  the  University  of
 * Tennessee, Knoxville, Innovative Computing Laboratory.             
 *                                                                      
 * 4. The name of the  University,  the name of the  Laboratory,  or the
 * names  of  its  contributors  may  not  be used to endorse or promote
 * products  derived   from   this  software  without  specific  written
 * permission.                                                          
 *                                                                      
 * -- Disclaimer:                                                       
 *                                                                      
 * THIS  SOFTWARE  IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,  INCLUDING,  BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY
 * OR  CONTRIBUTORS  BE  LIABLE FOR ANY  DIRECT,  INDIRECT,  INCIDENTAL,
 * SPECIAL,  EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES  (INCLUDING,  BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT,  STRICT LIABILITY,  OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 * ---------------------------------------------------------------------
 */ 
/*
 * Include files
 */
#include "hpl.h"

#define   I_SEND    0
#define   I_RECV    1

#ifdef STDC_HEADERS
void HPL_rollT
(
   HPL_T_panel *                    PBCST,
   int *                            IFLAG,
   HPL_T_panel *                    PANEL,
   const int                        N,
   double *                         U,
   const int                        LDU,
   const int *                      IPLEN,
   const int *                      IPMAP,
   const int *                      IPMAPM1
)
#else
void HPL_rollT
( PBCST, IFLAG, PANEL, N, U, LDU, IPLEN, IPMAP, IPMAPM1 )
   HPL_T_panel *                    PBCST;
   int *                            IFLAG;
   HPL_T_panel *                    PANEL;
   const int                        N;
   double *                         U;
   const int                        LDU;
   const int *                      IPLEN;
   const int *                      IPMAP;
   const int *                      IPMAPM1;
#endif
{
/* 
 * Purpose
 * =======
 *
 * HPL_rollT rolls the local arrays containing the local pieces of U, so
 * that on exit to this function  U  is replicated in every process row.
 * In addition, this function probe for the presence of the column panel
 * and forwards it when available.
 *
 * Arguments
 * =========
 *
 * PBCST   (local input/output)          HPL_T_panel *
 *         On entry,  PBCST  points to the data structure containing the
 *         panel (to be broadcast) information.
 *
 * IFLAG   (local input/output)          int *
 *         On entry, IFLAG  indicates  whether or not  the broadcast has
 *         already been completed.  If not,  probing will occur, and the
 *         outcome will be contained in IFLAG on exit.
 *
 * PANEL   (local input/output)          HPL_T_panel *
 *         On entry,  PANEL  points to the data structure containing the
 *         panel (to be rolled) information.
 *
 * N       (local input)                 const int
 *         On entry, N specifies the local number of rows of  U.  N must
 *         be at least zero.
 *
 * U       (local input/output)          double *
 *         On entry,  U  is an array of dimension (LDU,*) containing the
 *         local pieces of U in each process row.
 *
 * LDU     (local input)                 const int
 *         On entry, LDU specifies the local leading dimension of U. LDU
 *         should be at least  MAX(1,N).
 *
 * IPLEN   (global input)                const int *
 *         On entry, IPLEN is an array of dimension NPROW+1.  This array
 *         is such that IPLEN[i+1] - IPLEN[i] is the number of rows of U
 *         in each process row.
 *
 * IPMAP   (global input)                const int *
 *         On entry, IMAP  is an array of dimension  NPROW.  This  array
 *         contains  the  logarithmic mapping of the processes. In other
 *         words,  IMAP[myrow]  is the absolute coordinate of the sorted
 *         process.
 *
 * IPMAPM1 (global input)                const int *
 *         On entry,  IMAPM1  is an array of dimension NPROW. This array
 *         contains  the inverse of the logarithmic mapping contained in
 *         IMAP: For i in [0.. NPROW) IMAPM1[IMAP[i]] = i.
 *
 * ---------------------------------------------------------------------
 */ 
/*
 * .. Local Variables ..
 */
#if 0
   MPI_Datatype               type[2];
#endif
   MPI_Status                 status;
   MPI_Request                request;
   MPI_Comm                   comm;
   int                        Cmsgid=MSGID_BEGIN_PFACT, ibufR, ibufS,
                              ierr=MPI_SUCCESS, il, k, l, lengthR, 
                              lengthS, mydist, myrow, next, npm1, nprow,
                              partner, prev;
/* ..
 * .. Executable Statements ..
 */
   if( N <= 0 ) return;

   npm1 = ( nprow = PANEL->grid->nprow ) - 1; myrow = PANEL->grid->myrow;
   comm = PANEL->grid->col_comm;
/*
 * Rolling phase
 */
   mydist = IPMAPM1[myrow];
   prev   = IPMAP[MModSub1( mydist, nprow )];
   next   = IPMAP[MModAdd1( mydist, nprow )];
 
   for( k = 0; k < npm1; k++ )
   {
      l = (int)( (unsigned int)(k) >> 1 );
 
      if( ( ( mydist + k ) & 1 ) != 0 )
      {
         il      = MModAdd( mydist, l,   nprow );
         lengthS = IPLEN[il+1] - ( ibufS = IPLEN[il] );
         il    = MModSub( mydist, l+1, nprow );
         lengthR = IPLEN[il+1] - ( ibufR = IPLEN[il] ); partner = prev;
      }
      else
      {
         il    = MModSub( mydist, l,   nprow );
         lengthS = IPLEN[il+1] - ( ibufS = IPLEN[il] );
         il    = MModAdd( mydist, l+1, nprow );
         lengthR = IPLEN[il+1] - ( ibufR = IPLEN[il] ); partner = next;
      }
 
      if( lengthR > 0 )
      {
#if 0
         if( ierr == MPI_SUCCESS )
         {
            if( LDU == N )
               ierr = MPI_Type_contiguous( lengthR * LDU, MPI_DOUBLE,
                                           &type[I_RECV] );
            else
               ierr = MPI_Type_vector( lengthR, N, LDU, MPI_DOUBLE,
                                       &type[I_RECV] );
         }
         if( ierr == MPI_SUCCESS )
            ierr =   MPI_Type_commit( &type[I_RECV] );
         if( ierr == MPI_SUCCESS )
            ierr =   MPI_Irecv( Mptr( U, 0, ibufR, LDU ), 1, type[I_RECV],
                                partner, Cmsgid, comm, &request );
#else
/*
 * In our case, LDU is N - Do not use the MPI datatype.
 */
         if( ierr == MPI_SUCCESS )
            ierr =   MPI_Irecv( Mptr( U, 0, ibufR, LDU ), lengthR*LDU,
                                MPI_DOUBLE, partner, Cmsgid, comm, &request );
#endif
      }
 
      if( lengthS > 0 )
      {
#if 0
         if( ierr == MPI_SUCCESS )
         {
            if( LDU == N )
               ierr =   MPI_Type_contiguous( lengthS*LDU, MPI_DOUBLE,
                                             &type[I_SEND] );
            else
               ierr =   MPI_Type_vector( lengthS, N, LDU, MPI_DOUBLE,
                                         &type[I_SEND] );
         }
         if( ierr == MPI_SUCCESS )
            ierr =   MPI_Type_commit( &type[I_SEND] );
         if( ierr == MPI_SUCCESS )
            ierr =   MPI_Send( Mptr( U, 0, ibufS, LDU ), 1, type[I_SEND],
                               partner, Cmsgid, comm );
         if( ierr == MPI_SUCCESS )
            ierr =   MPI_Type_free( &type[I_SEND] );
#else
/*
 * In our case, LDU is N - Do not use the MPI datatype.
 */
         if( ierr == MPI_SUCCESS )
            ierr =   MPI_Send( Mptr( U, 0, ibufS, LDU ), lengthS*LDU,
                               MPI_DOUBLE, partner, Cmsgid, comm );
#endif
      }

      if( lengthR > 0 )
      {
         if( ierr == MPI_SUCCESS )
            ierr =   MPI_Wait( &request, &status );
#if 0
         if( ierr == MPI_SUCCESS )
            ierr =   MPI_Type_free( &type[I_RECV] );
#endif
      }
/*
 * Probe for column panel - forward it when available
 */
      if( *IFLAG == HPL_KEEP_TESTING ) (void) HPL_bcast( PBCST, IFLAG );
   }

   if( ierr != MPI_SUCCESS )
   { HPL_pabort( __LINE__, "HPL_rollT", "MPI call failed" ); }
/*
 * End of HPL_rollT
 */
}
